File: Project/MyProject/MyProject/Source/MyProject/Private/BasicFrame/Libs/FrameHandle/TickObjectPriorityMgr.cpp
#include "MyProject.h"
#include "TickObjectPriorityMgr.h"
#include "MClassFactory.h"
#include "MacroDef.h"
#include "Ctx.h"
#include "LogSys.h"
#include "LogTypeId.h"
#include "INoOrPriorityList.h"
#include "GObject.h"

MY_BEGIN_NAMESPACE(MyNS)

M_IMPLEMENT_AND_REGISTER_CLASS(TickObjectPriorityMgr, DelayPriorityHandleMgr)

TickObjectPriorityMgr::TickObjectPriorityMgr()
{

}

void TickObjectPriorityMgr::init()
{
	Super::init();
}

void TickObjectPriorityMgr::dispose()
{
	Super::dispose();
}

void TickObjectPriorityMgr::onTick(float delta, TickMode tickMode)
{
	this->advance(delta, tickMode);
}

void TickObjectPriorityMgr::advance(float delta, TickMode tickMode)
{
	this->_incDepth();

	this->_onPreAdvance(delta, tickMode);
	this->_onExecAdvance(delta, tickMode);
	this->_onPostAdvance(delta, tickMode);

	this->_decDepth();
}

void TickObjectPriorityMgr::_onPreAdvance(float delta, TickMode tickMode)
{

}

void TickObjectPriorityMgr::_onExecAdvance(float delta, TickMode tickMode)
{
	int index = 0;
	int count = this->mNoOrPriorityList->count();
	ITickedObject* tickObject = nullptr;
	GObject* gObject = nullptr;

	while (index < count)
	{
		tickObject = (ITickedObject*)this->mNoOrPriorityList->get(index);

		if (nullptr != (IDelayHandleItem*)tickObject)
		{
			//if (!((IDelayHandleItem*)tickObject)->isClientDispose())

			gObject = (GObject*)tickObject;

			if (!gObject->isClientDispose())
			{
				gObject->onTick(delta, tickMode);
			}
		}
		else
		{
			if (MacroDef::ENABLE_LOG)
			{
				GLogSys->log("TickObjectPriorityMgr::_onExecAdvance, failed", LogTypeId::eLogCommon);
			}
		}

		++index;
	}
}

void TickObjectPriorityMgr::_onPostAdvance(float delta, TickMode tickMode)
{

}

MY_END_NAMESPACE

File: Project/MyProject/Tool/CSToolKit/Project/ToolKit/ToolKit/Source/Libs/FrameHandle/TickObjectPriorityMgr.cs
namespace SDK.Lib
{
    /**
     * @brief 每一帧执行的对象管理器
     */
    public class TickObjectPriorityMgr : DelayPriorityHandleMgr, ITickedObject, INoOrPriorityObject
    {
        public TickObjectPriorityMgr()
        {

        }

        override public void init()
        {
            base.init();
        }

        override public void dispose()
        {
            base.dispose();
        }

        public void onTick(float delta, TickMode tickMode)
        {
            this.advance(delta, tickMode);
        }

        virtual public void advance(float delta, TickMode tickMode)
        {
            this._incDepth();

            this._onPreAdvance(delta, tickMode);
            this._onExecAdvance(delta, tickMode);
            this._onPostAdvance(delta, tickMode);

            this._decDepth();
        }

        virtual protected void _onPreAdvance(float delta, TickMode tickMode)
        {

        }

        virtual protected void _onExecAdvance(float delta, TickMode tickMode)
        {
            int idx = 0;
            int count = this.mNoOrPriorityList.count();
            ITickedObject tickObject = null;

            while (idx < count)
            {
                tickObject = this.mNoOrPriorityList.get(idx) as ITickedObject;

                if (null != (tickObject as IDelayHandleItem))
                {
                    if (!(tickObject as IDelayHandleItem).isClientDispose())
                    {
                        tickObject.onTick(delta, tickMode);
                    }
                }
                else
                {
                    if (MacroDef.ENABLE_LOG)
                    {
                        Ctx.msInstance.mLogSys.log("TickObjectPriorityMgr::_onExecAdvance, failed", LogTypeId.eLogCommon);
                    }
                }

                ++idx;
            }
        }

        virtual protected void _onPostAdvance(float delta, TickMode tickMode)
        {

        }
    }
}